protocols/kad/src/kbucket/bucket.rs

```diff
@@ -0,0 +1,552 @@
+pub const MAX_NODES_PER_BUCKET: usize = 20;
+
+/// A `PendingNode` is a `Node` that is pending insertion into a `KBucket`.
+#[derive(Debug, Clone)]
+pub struct PendingNode<TPeerId, TVal> {
+    /// The pending node to insert.
+    node: Node<TPeerId, TVal>,
+
+    /// The status of the pending node.
+    status: NodeStatus,
+
+    /// The instant at which the pending node is eligible for insertion into a bucket.
+    replace: Instant,
+}
+
+/// The status of a node in a bucket.
+///
+/// The status of a node in a bucket together with the time of the
+/// last status change determines the position of the node in a
+/// bucket.
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum NodeStatus {
+    /// The node is considered connected.
+    Connected,
+    /// The node is considered disconnected.
+    Disconnected
+}
+
+impl<TPeerId, TVal> PendingNode<TPeerId, TVal> {
+    pub fn key(&self) -> &Key<TPeerId> {
+        &self.node.key
+    }
+
+    pub fn status(&self) -> NodeStatus {
+        self.status
+    }
+
+    pub fn value_mut(&mut self) -> &mut TVal {
+        &mut self.node.value
+    }
+
+    pub fn is_ready(&self) -> bool {
+        Instant::now() >= self.replace
+    }
+
+    pub fn set_ready_at(&mut self, t: Instant) {
+        self.replace = t;
+    }
+}
+
+/// A `Node` in a bucket, representing a peer participating
+/// in the Kademlia DHT together with an associated value (e.g. contact
+/// information).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Node<TPeerId, TVal> {
+    /// The key of the node, identifying the peer.
+    pub key: Key<TPeerId>,
+    /// The associated value.
+    pub value: TVal,
+}
+
+/// The position of a node in a `KBucket`, i.e. a non-negative integer
+/// in the range `[0, MAX_NODES_PER_BUCKET)`.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Position(usize);
+
+/// A `KBucket` is a list of up to `MAX_NODES_PER_BUCKET` `Key`s and associated values,
+/// ordered from least-recently connected to most-recently connected.
+#[derive(Debug, Clone)]
+pub struct KBucket<TPeerId, TVal> {
+    /// The nodes contained in the bucket.
+    nodes: ArrayVec<[Node<TPeerId, TVal>; MAX_NODES_PER_BUCKET]>,
+
+    /// The position (index) in `nodes` that marks the first connected node.
+    ///
+    /// Since the entries in `nodes` are ordered from least-recently connected to
+    /// most-recently connected, all entries above this index are also considered
+    /// connected, i.e. the range `[0, first_connected_pos)` marks the sub-list of entries
+    /// that are considered disconnected and the range
+    /// `[first_connected_pos, MAX_NODES_PER_BUCKET)` marks sub-list of entries that are
+    /// considered connected.
+    ///
+    /// `None` indicates that there are no connected entries in the bucket, i.e.
+    /// the bucket is either empty, or contains only entries for peers that are
+    /// considered disconnected.
+    first_connected_pos: Option<usize>,
```

Comment:
I find the semantics clearer if `first_connected_pos` is either not set (`None`) or is a valid index into `nodes`, instead of attaching special meaning to out-of-bounds values (w.r.t. `nodes`).